compiler/src/iree/compiler/DispatchCreation/BubbleUpExtractSlices.cpp

```diff
@@ -0,0 +1,131 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/LinalgExt/Utils/Utils.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/Linalg/Utils/Utils.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Tensor/Transforms/Transforms.h"
+#include "mlir/Dialect/Utils/ReshapeOpsUtils.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir::iree_compiler::DispatchCreation {
+
+#define GEN_PASS_DEF_BUBBLEUPEXTRACTSLICESPASS
+#include "iree/compiler/DispatchCreation/Passes.h.inc"
+
+namespace {
+
+// Convert extract_slice(dequant) to dequant(extract_slice)
+//
+// Because `extract_slice` ops and dequantize-like ops get cloned into regions
+// later, its okay to bubble up through multi-use dequant ops.
+struct BubbleUpExtract : OpRewritePattern<tensor::ExtractSliceOp> {
+  using OpRewritePattern<tensor::ExtractSliceOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::ExtractSliceOp sliceOp,
+                                PatternRewriter &rewriter) const final {
+    Value source = sliceOp.getSource();
+    auto linalgOp = source.getDefiningOp<linalg::LinalgOp>();
+    if (!linalgOp || linalgOp->getNumResults() != 1) {
+      return rewriter.notifyMatchFailure(
+          sliceOp, "expected source to implement `linalg::LinalgOp` and have a "
+                   "single result");
+    }
+
+    if (!IREE::LinalgExt::isBitExtendOp(linalgOp)) {
+      return rewriter.notifyMatchFailure(
+          sliceOp, "expected source to be dequantize-like");
+    }
+
+    if (!sliceOp.hasUnitStride()) {
+      return rewriter.notifyMatchFailure(sliceOp, "expected unit stride");
+    }
+
+    Value replacement;
+    linalg::GenericOp swappedOp;
+    {
+      FailureOr<TilingResult> tilingResult =
+          tensor::replaceExtractSliceWithTiledProducer(rewriter, sliceOp,
+                                                       linalgOp->getResult(0));
+      if (failed(tilingResult)) {
+        return rewriter.notifyMatchFailure(
+            linalgOp, "failed to swap extract_slice with op");
+      }
+      if (tilingResult->tiledOps.size() != 1 ||
+          !isa<linalg::GenericOp>(tilingResult->tiledOps[0])) {
+        return rewriter.notifyMatchFailure(
+            linalgOp, "expected extract_slice to generate a `linalg.generic`");
```

Comment:
Good point, I removed this check and added the checks you suggested before IR mutation. I also changed the `failed` checks so asserts since there is no graceful way to exit at that point